<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML><HEAD><TITLE>Reading Command Console Data Files</TITLE>
<META http-equiv=Content-Type content="text/html; charset=windows-1252">
<META 
content="Affymetrix, GeneChip, GeneChip Array, Netaffx, NetAffx Analysis Center, Affymetrics, Affimetrix, Affimetrics, " 
name=keywords>
<META 
content="Affymetrix is dedicated to developing state-of-the-art technology for acquiring, analyzing, and managing complex genetic information for use in biomedical research. Affymetrix sells GeneChip® brand arrays." 
name=description>
<SCRIPT language=JavaScript>
</SCRIPT>

<SCRIPT language=JavaScript src="file_formats/files_files/main.js" 
type=text/javascript></SCRIPT>
<LINK href="file_formats/files_files/style.css" type=text/css rel=stylesheet><LINK 
href="/favicon.ico" rel=icon>
<META content="Microsoft FrontPage 6.0" name=GENERATOR>
<style>
</style>
</HEAD>
<BODY onresize=fixIt() vLink=#003366 aLink=#003366 link=#003366 bgColor=#ffffff 
leftMargin=0 topMargin=0 onunload=window.focus() marginheight="0" 
marginwidth="0">

<NOINDEX>

<TABLE cellSpacing=0 cellPadding=0 border=0> 
<TBODY>
<TR>
	<TD vAlign=center align=left>&nbsp;</TD>
</TR>
</TBODY>
</TABLE>

<TABLE cellSpacing=0 cellPadding=0 border=0 width="601"> 
<TBODY>
<TR>
	<TD vAlign=top align=left><IMG height=54 alt=Affymetrix src="file_formats/files_files/affymetrix_logo.gif" width=128 border=0></TD>
	<TD></TD>
	<TD vAlign=center align=left class=pageheader><p>AFFYMETRIX DEVELOPER NETWORK</p></TD>
</TR>
</TBODY>
</TABLE>

</NOINDEX>

<TABLE CELLSPACING=0 CELLPADDING=0 WIDTH=772 BORDER=0>
<TBODY>
<TR>
	<TD VALIGN=top ALIGN=left WIDTH=772 BGCOLOR=#336699 HEIGHT="2"><IMG height=3 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</TR> 
</TBODY>
</TABLE>

<TABLE cellSpacing=0 cellPadding=5 width=409 background=file_formats/files_files/network_background.gif border=0>
<TBODY>
<TR>
	<TD vAlign=top align=left> 
	<TABLE cellSpacing=0 cellPadding=0 background=file_formats/files_files/spacer.gif border=0>
	<TBODY>
	<TR>
	    <TD vAlign=top align=left width=11" ><IMG HEIGHT=1 ALT="" SRC="file_formats/files_files/spacer.gif" WIDTH=11 BORDER=0></TD>
		<TD vAlign=top align=left width="100%"><IMG height=41 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
	</TR>
	<TR> 
		<TD vAlign=top align=left width=11" ><IMG HEIGHT=1 ALT="" SRC="file_formats/files_files/spacer.gif" WIDTH=11 BORDER=0></TD>
		<TD class=intro vAlign=top align=left><P><span class="header2">Reading 
		Command Console Data Files</span></P></TD>
	</TR>
	</TBODY>
	</TABLE>
	</TD>
</TR>
</TBODY>
</TABLE>

<br>

<TABLE cellSpacing=0 cellPadding=0 border=0 class="content"> <TBODY>
<tr>
	<TD>&nbsp;</TD>
</tr>
<tr>
	<TD width="17" align="right"><IMG height=11 alt="" src="file_formats/files_files/bullet2.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></TD>
	<TD class=header3>INTRODUCTION</TD>
</tr>
<tr>
	<TD><IMG height=10 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</tr>
<tr>
	<TD width="17"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></P></TD>
	<TD class=content">
		The Fusion SDK provides source code for reading both GCOS 1.x and 
		Command Console data files. The way the SDK is organized is depicted in 
		the diagram below.<p><img border="0" src="fusion_layers.gif"></p>
		<p>The GCOS files parsers are file specific as each data file has its 
		own format.</p>
		<p>The Command Console (CC) Array file is an XML file. The XML parser 
		used is the Apache Xerces software.</p>
		<p>The CC data files use the same flexible file format. This format 
		contains information to describe the contents (#columns, data type in 
		each column, etc.). A set of classes provide the ability to read any 
		data file of this format, regardless of file extension. For existing 
		files such as the CEL and CHP files which contain results generated by 
		the current GCOS/GTYPE/GDAS software, file specific interpreters provide 
		higher level interfaces. These interpreters make assumptions about the 
		number of data groups, sets and column which the CC software will 
		conform to. For assays that will write CHP files with newer data (such 
		as the Tiling and ExACT software) the fusion layer will utilize the 
		generic file parsers.</p>
		<p>The information below will describe how to read each data file type 
		using the Fusion APIs.</td>
</tr>
<tr>
	<TD><IMG height=10 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</tr>
<tr>
	<TD width="17" align="right"><IMG height=11 alt="" src="file_formats/files_files/bullet2.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></TD>
	<TD class=header3>ARRAY FILE</TD>
</tr>
<tr>
	<TD><IMG height=10 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</tr>
<tr>
	<TD width="17"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></P></TD>
	<TD class=content">
		The CC array file contains information about a GeneChip probe array and 
		end user attributes.<p>The top level Fusion objects to use are <i><b>
		FusionArrayFileReader</b></i> and <b><i>ArrayData</i></b>. These classes 
		will parse CC array files, MAGE-ML files generated by the DTT 1.1 
		software and EXP files.</p>
		<p>The workflow for reading the array file is:</p>
		<ol>
			<li>Create a FusionArrayFileReader object.</li>
			<li>Read the file using the Read() function passing in the full path 
			name of the file and the ArrayData object.</li>
		</ol>
	</td>
</tr>
<tr>
	<TD><IMG height=10 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</tr>
<tr>
	<TD width="17" align="right"><IMG height=11 alt="" src="file_formats/files_files/bullet2.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></TD>
	<TD class=header3>CEL FILE</TD>
</tr>
<tr>
	<TD><IMG height=10 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</tr>
<tr>
	<TD width="17"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></P></TD>
	<TD class=content">
		The CEL file contains intensity data organized by X/Y feature 
		coordinate. Included in the file is the name of the algorithm and the 
		parameters used to create the file.<p>The top level Fusion object to use 
		is <i><b>FusionCELData</b></i>.</p>
		<p>The workflow for reading a CEL file is:</p>
		<ol>
			<li>Create a FusionCELData object.</li>
			<li>Set the full path name of the input CEL file using the 
			SetFileName() function.</li>
			<li>Read the file using the Read() function. A false boolean value 
			may be passed into the function if mask and outlier data are not to 
			be read.</li>
			<li>Use the GetParameters() function to retrieve the list of 
			algorithm parameters.</li>
			<li>Use the GetAlg() function to retrieve the name of the algorithm 
			used to create the CEL file.</li>
			<li>Various function are available to retrieve the intensity data. 
			These include GetEntry() to retrieve the intensity, standard 
			deviation and pixel count for a given feature and GetIntensity() to 
			retrieve just the intensity value.</li>
		</ol>
		<p>Note: Use the CDF file to organize the data by probe set.</tr>
</tr>
<tr>
	<TD><IMG height=10 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</tr>
<tr>
	<TD width="17" align="right"><IMG height=11 alt="" src="file_formats/files_files/bullet2.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></TD>
	<TD class=header3>CHP FILE</TD>
</tr>
<tr>
	<TD><IMG height=10 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</tr>
<tr>
	<TD width="17"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></P></TD>
	<TD class=content">
		The CHP file contains the results from analyze intensity data in a CEL 
		file. For expression arrays this is the probe set results such as signal 
		and detection. For genotyping this is the allele calls and confidence 
		scores.<p>Reading a CHP file involves several Fusion classes. To read 
		the file use the <i><b>FusionCHPData</b></i> and <i><b>FusionCHPDataReg</b></i> 
		classes.</p>
		<blockquote>
			<p><i>FusionCHPData *chp = FusionCHPDataReg::Read(fileName);&nbsp; 
			// fileName is the name of the input CHP file.</i></p>
		</blockquote>
		<p>To support the multitude of expected CHP file types, the CHP parser 
		has been written using self registering classes with each self 
		registered class providing the parsing capabilities for the specific CHP 
		file type. The existing CHP parsers include:</p>
		<ul>
			<li>FusionCHPLegacyData - This class provides parsing for CHP files 
			containing results from our legacy assays (expression, genotyping, 
			resequencing, tag).</li>
			<li>FusionCHPTilingData - This class provides parsing for CHP files 
			containing results from the Tiling Analysis Software.</li>
			<li>FusionCHPGenericData - This class provides a catch all for CHP 
			files where a data specific interpreter is not provided. This class 
			directly accesses the generic data file parser and data objects, no 
			assumptions are made as to how many data groups, sets and columns 
			are stored in the file.</li>
			<li>FusionCHPQuantificationData - This class provides parsing for 
			CHP files containing results from the RMA or PLIER algorithms from 
			the Expression Console software for 3' Expression arrays.</li>
			<li>FusionCHPQuantificationDetectionData - This class provides 
			parsing for CHP files containing results from the Expression Console 
			software for Exon arrays.</li>
			<li>FusionCHPMultiDataData - This class provides parsing for CHP 
			files containing genotyping analysis results from the BRLMM Analysis 
			Tool Version 2 or Genotyping Console™ software.</li>
		</ul>
		<p>Each of the above classes contains a function (FromBase) to convert 
		the FusionCHPData object. An example of converting to the legacy class 
		is:</p>
		<blockquote>
			<p><i>FusionCHPLegacyData *legacychp = 
			FusionCHPLegacyData::FromBase(chp);</i></p>
		</blockquote>
		<p>The return object is NULL if the file is not an expression, 
		genotyping, resequencing or tag CHP file (those data types supported by 
		the FusionCHPLegacyData object).</p>
		<p>At this point you need to decide what to do. If your application 
		reads expression data and you want to extract the signal values for each 
		probe set then the workflow steps are:</p>
		<ol>
			<li>Verify that the returned pointer from the FromBase() function is 
			not NULL.</li>
			<li>Verify that the data in the CHP file is expression using the 
			GetHeader().GetAssayType() function. The value should be &quot;FusionExpression&quot;.</li>
			<li>Loop over the number of probe sets - defined by the 
			GetHeader().GetNumProbeSets() function.</li>
			<li>Retrieve the probe set object using the GetExpressionResults() 
			function. This returns a FusionExpressionProbeSetResults object.</li>
			<li>The signal value is retrieved using the 
			FusionExpressionProbeSetResults.GetSignal() function.</li>
		</ol>
		<p>Note: As of GCOS 1.0 the CEL intensities are no stored in the CHP 
		file. Use the CEL parser to retrieve intensity information.</tr>
</tr>
<tr>
	<TD><IMG height=10 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</tr>
<tr>
	<TD width="17" align="right"><IMG height=11 alt="" src="file_formats/files_files/bullet2.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></TD>
	<TD class=header3>CDF FILE</TD>
</tr>
<tr>
	<TD><IMG height=10 alt="" src="file_formats/files_files/spacer.gif" width=1 border=0></TD>
</tr>
<tr>
	<TD width="17"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=11 border=0></TD>
	<TD width="7"><IMG height=1 alt="" src="file_formats/files_files/spacer.gif" width=5 border=0></P></TD>
	<TD class=content">
		The CDF file provides a description of the array in terms of function of 
		the probes.<p>The top level Fusion object to use is <b><i>FusionCDFData</i></b>.</p>
		<p>The workflow for reading a CDF file is:</p>
		<ol>
			<li>Create a FusionCDFData object.</li>
			<li>Set the full path name of the input CDF file using the 
			SetFileName() function. The name of the file is the array type + 
			&quot;.CDF&quot;. The array type may be extracted from a CEL file using the 
			FusionCELData.GetChipType() function.</li>
			<li>Read the file using the Read() function.</li>
			<li>The name of a probe set may be extracted using the 
			GetProbeSetName() function.</li>
			<li>Loop over the number of probe sets. The number of probe sets is 
			determined using the GetHeader().GetNumProbeSets() function. The 
			probe set information is retrieved in a FusionCDFProbeSetInformation 
			object.</li>
			<li>The probe set contains one or more groups (1 for expression, 
			typically 2 or 4 for genotyping). The number of probe groups is 
			determined using the GetNumGroups() function. Call the 
			GetGroupInformation() function to retrieve the group information.</li>
			<li>The group contains one ore more probes, defined by the 
			GetNumCells() function. Each probe information may be extracted 
			using the GetCell() function.</li>
		</ol>
		<p>Once you have a probe object you can determine if it is a PM or MM 
		probe by examining the PBase and TBase. PM probes have PBase equal to 
		the compliment of TBase, otherwise it is a MM probe.</tr>
</tr>
  
	
</TABLE></TD></TR></TBODY>
<p><BR> 
</p>
</BODY></HTML>